### PR TITLE
Implement loglogistic distribution

### PR DESCRIPTION
**Pull request by npfp**
Formula from the
Handbook of Differential Entropy, J.V. Michalowicsz, J.M. Nichols, F. Bucholtz
p. 113, 114

### CODE REVIEW

File: tensorflow_probability/python/distributions/loglogistic.py [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+__all__ = [
+    'LogLogistic',
+]
+
+
+class LogLogistic(transformed_distribution.TransformedDistribution):
+  """The log-logistic distribution."""
+
+  def __init__(self,
+               alpha,
```
Comment (brianwa84) [inline tensorflow_probability/python/distributions/loglogistic.py]: We typically use terms like `scale` and `concentration` for these, and avoid greek.

**Comment by srvasude** [comment]
+1 I might try to rephrase this as a TransformedDistribution.

**Comment by npfp** [comment]
Yes, the main point is to use the (relatively simple) closed formula for the various quantities and gain some numerical stability.

File: tensorflow_probability/python/distributions/loglogistic.py (hunk 1 above)
Comment (npfp) [reply tensorflow_probability/python/distributions/loglogistic.py]: Regarding the greeks, I'll change that, thx

**Comment by npfp** [comment]
@srvasude I'm not sure to understand. It's already inheriting from TransformedDistribution

**Comment by srvasude** [comment]
Sorry I meant to use as much as TransformedDistribution as you can. But you already answered why you are overriding methods, so you can ignore my comment :).

**Comment by npfp** [comment]
Hi there, Hope you're safe.
 I just wonder if there was anything pending on my side on this PR?

File: tensorflow_probability/python/distributions/loglogistic.py [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+    return self._concentration
+
+  def _mean(self):
+    b = 1. / self.concentration
+    mean = self.scale / sinc(b)
+    return tf.where(tf.greater(self.concentration, 1.), mean, np.nan)
+
+  def _variance(self):
+    b = 1. / self.concentration
+    variance = self.scale ** 2 * (1. / sinc(2*b) - 1. / sinc(b)**2)
```
Comment (srvasude) [inline tensorflow_probability/python/distributions/loglogistic.py:108]: Do you know how numerically stable this calculation is?

File: tensorflow_probability/python/distributions/loglogistic.py [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+    return tf.where(tf.greater(self.concentration, 2.), variance, np.nan)
+
+  def _mode(self):
+    mode = self.scale * ((self.concentration - 1.)/(self.concentration + 1.)
```
Comment (srvasude) [inline tensorflow_probability/python/distributions/loglogistic.py:112]: Can this be done in log space (using `expm1` and `log1p` for better numerics)?

File: tensorflow_probability/python/distributions/loglogistic.py [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+                         )**(1./self.concentration)
+    return tf.where(tf.greater(self.concentration, 1.), mode, np.nan)
+
+  def _entropy(self):
+    return (tf.math.log(self.scale / self.concentration) + 2.) / tf.math.log(2.)
```
Comment (srvasude) [inline tensorflow_probability/python/distributions/loglogistic.py:117]: Prefer `np.log(2)`

File: tensorflow_probability/python/distributions/loglogistic_test.py [hunk 5]
```diff
@@ -0,0 +1,110 @@
+# Copyright 2018 The TensorFlow Probability Authors.
```
Comment (srvasude) [inline tensorflow_probability/python/distributions/loglogistic_test.py:1]: 2020

File: tensorflow_probability/python/distributions/loglogistic.py [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+               scale,
+               concentration,
+               validate_args=False,
+               allow_nan_stats=True,
+               name='LogLogistic'):
+    """Construct a log-logistic distribution.
+
+    The LogLogistic distribution models positive-valued random variables
+    whose logarithm is logisticly distributed with mean `log(scale)` and
+    scale `1/concentration`. It is constructed as the exponential
```
Comment (srvasude) [inline tensorflow_probability/python/distributions/loglogistic.py:49]:
```suggestion
    scale `1 / concentration`. It is constructed as the exponential
```

File: tensorflow_probability/python/distributions/loglogistic_test.py [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+from tensorflow_probability.python.internal import test_util
+
+
+@test_util.test_all_tf_execution_regimes
+class LogLogiticTest(test_util.TestCase):
+
+  def setUp(self):
+    self._rng = np.random.RandomState(123)
+
+  def testLogLogisticStats(self):
```
Comment (srvasude) [inline tensorflow_probability/python/distributions/loglogistic_test.py:35]: Separate out tests?

File: tensorflow_probability/python/distributions/loglogistic.py (hunk 2 above)
Comment (npfp) [reply tensorflow_probability/python/distributions/loglogistic.py]: I'm having a look

File: tensorflow_probability/python/distributions/loglogistic.py (hunk 3 above)
Comment (npfp) [reply tensorflow_probability/python/distributions/loglogistic.py]: I'm having a look

**Comment by npfp** [comment]
@googlebot I fixed it.

File: tensorflow_probability/python/distributions/loglogistic.py (hunk 3 above)
Comment (npfp) [reply tensorflow_probability/python/distributions/loglogistic.py]: I'm not sure it will improve numerical stability:
it's only computed for `c >1`, so

```
[(c-1)/(c+1)]**(1/c) = exp{[ln(c-1) - ln(c+1)]/c}
```

The numerical issue is when `c~1` where `exp{- ln(c+1)/c} ~1/2` and therefore won't benefit from `log1p`.

What do you think ?

File: tensorflow_probability/python/distributions/loglogistic.py (hunk 2 above)
Comment (npfp) [reply tensorflow_probability/python/distributions/loglogistic.py]: There will be some numerical issues when `c -> 2`, and the only way to solve them would be to work out a numerical approximation  when `c~2` and use it with something like:

```
tf.where(c < 2.+ tol, approximation, real_value)
```
But I'm no sure if it would be a good thing to introduce a `tf.where` and a `tol`. What do you think ?

File: tensorflow_probability/python/distributions/loglogistic.py [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+    with tf.name_scope(name) as name:
+      dtype = dtype_util.common_dtype([scale, concentration],
+                                      dtype_hint=tf.float32)
+      self._scale = tensor_util.convert_nonref_to_tensor(
+          scale, name='scale', dtype=dtype)
+      self._concentration = tensor_util.convert_nonref_to_tensor(
+          concentration, name='concentration', dtype=dtype)
+      super(LogLogistic, self).__init__(
+          distribution=logistic.Logistic(loc=tf.math.log(self.scale),
+                                         scale=1./self.concentration),
```
Comment (srvasude) [inline tensorflow_probability/python/distributions/loglogistic.py:76]:
```suggestion
                                         scale=tf.math.reciprocal(self.concentration)),
```

File: tensorflow_probability/python/distributions/loglogistic.py [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+    parameters = dict(locals())
+    with tf.name_scope(name) as name:
+      dtype = dtype_util.common_dtype([scale, concentration],
+                                      dtype_hint=tf.float32)
+      self._scale = tensor_util.convert_nonref_to_tensor(
+          scale, name='scale', dtype=dtype)
+      self._concentration = tensor_util.convert_nonref_to_tensor(
+          concentration, name='concentration', dtype=dtype)
+      super(LogLogistic, self).__init__(
+          distribution=logistic.Logistic(loc=tf.math.log(self.scale),
```
Comment (srvasude) [inline tensorflow_probability/python/distributions/loglogistic.py:75]: This can be problematic (this and below). This is because if these are `tf.Variable`, this won't be gradient tape safe.

I wonder if it's better to accept `loc` and `scale` as parameters. We do this for the `LogNormal` distribution and use the parametrization, and so would be consistent with it. WDYT?

Comment (npfp) [reply tensorflow_probability/python/distributions/loglogistic.py:75]: Why is it not gradient tape safe? Is it because the operations are defined at the initialisation ?

Regarding your suggestion, I'm not sure. I agree it would good to be consistent with the `LogNormal` but in the meantime, by doing so we lose the possibility to introduce a `loc` parameter later on  (and it makes me realise that if we keep this option I should remove:

```
  @property
  def loc(self):
    """Distribution parameter for the pre-transformed mean."""
    return self.distribution.loc
```

File: tensorflow_probability/python/distributions/loglogistic.py [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+                                         scale=1./self.concentration),
+          bijector=exp_bijector.Exp(),
```
[the remark this answers is not in the document]
Comment (npfp) [reply tensorflow_probability/python/distributions/loglogistic.py:80]: I implemented them as well as log_cdf and log_survival_function as both of them can also gain some numerical stability.

File: tensorflow_probability/python/distributions/loglogistic.py (hunk 9 above)
Comment (srvasude) [reply tensorflow_probability/python/distributions/loglogistic.py]: Yep that's exactly it.

I would probably try to be consistent with LogNormal. Also IIRC, GeneralizedLogLogistic has a slightly different parameterization so we might want to have that as a separate implementation anyways.

Comment (npfp) [reply tensorflow_probability/python/distributions/loglogistic.py]: I see, thx.

I juste pushed a change but for some reasons I got nans for a couple of numpy tests only. Would you have any clue on why the numpy tests behave differently than the others?

**Comment by npfp** [comment]
> You need to sign the CLA

I did sign it.

I believe the CLA check is failing because I committed @srvasude suggested changes. See https://github.com/tensorflow/probability/pull/843#issuecomment-620167926

**Comment by srvasude** [comment]
Hrmm this came up before. For some reason it is using an @gmail.com account. I think you can change my email for the suggestion to @google.com and things should work?

File: tensorflow_probability/python/distributions/loglogistic.py [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+
+  def _mean(self):
+    with tf.control_dependencies(
+        [] if self.allow_nan_stats else [  # pylint: disable=g-long-ternary
+          assert_util.assert_less(
+            self.scale,
+            tf.ones([], dtype=self.dtype),
+            message='Mean undefined for scale > 1.'),
+        ]):
+      mean = tf.math.exp(self.loc) / sinc(self.scale)
```
Comment (srvasude) [inline tensorflow_probability/python/distributions/loglogistic.py:101]: For passing distribution properties test you will need to do something like:

`scale = tf.convert_to_tensor(self.scale)`

and then use scale.

Multiple uses of a `self....` property will fail the test. This will need to be changed here and below

File: tensorflow_probability/python/distributions/loglogistic.py [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+            message='Variance undefined for scale > 1/2.'),
+        ]):
+      variance = tf.math.exp(2 * self.loc) * (
+          1. / sinc(2 * self.scale) - 1. / sinc(self.scale) ** 2)
+      return tf.where(self.scale > 0.5, np.nan, variance)
+
+  def _mode(self):
+    mode = tf.math.exp(
+      self.loc
+      + self.scale * (tf.math.log(1. - self.scale) - tf.math.log(self.scale)
```
Comment (srvasude) [inline tensorflow_probability/python/distributions/loglogistic.py:119]: `scale * (tf.math.log1p(-scale) - tf.math.log1p(scale))`

File: tensorflow_probability/python/distributions/loglogistic.py [hunk 13]
```diff
@@ -0,0 +1,161 @@
...
+            self.scale,
+            0.5 * tf.ones([], dtype=self.dtype),
+            message='Variance undefined for scale > 1/2.'),
+        ]):
+      variance = tf.math.exp(2 * self.loc) * (
+          1. / sinc(2 * self.scale) - 1. / sinc(self.scale) ** 2)
+      return tf.where(self.scale > 0.5, np.nan, variance)
+
+  def _mode(self):
+    mode = tf.math.exp(
```
Comment (srvasude) [inline tensorflow_probability/python/distributions/loglogistic.py:117]: Break this expression in to log_mode, and then exponentiate for the mode?

File: tensorflow_probability/python/distributions/loglogistic.py [hunk 14]
```diff
@@ -0,0 +1,161 @@
...
+  def _log_z(self, x):
+    """Returns log of the standardized input."""
+    with tf.name_scope('log_standardize'):
+      return (tf.math.log(x) - self.loc) / self.scale
+
+  def _log_prob(self, x):
+    return (-tf.math.log(self.scale) - self.loc
+            + (1. - self.scale) * self._log_z(x)
+            - 2 * tf.math.log1p(
+              tf.math.exp(self._log_z(x)))
```
Comment (srvasude) [inline tensorflow_probability/python/distributions/loglogistic.py:135]: Prefer using tf.math.softplus.

`tf.math.softplus(x) == tf.math.log1p(tf.math.exp(x))`

File: tensorflow_probability/python/distributions/loglogistic.py [hunk 15]
```diff
@@ -0,0 +1,161 @@
...
+            )
+
+  def _log_cdf(self, x):
+    return -tf.math.log1p(tf.math.exp(-self._log_z(x)))
```
Comment (srvasude) [inline tensorflow_probability/python/distributions/loglogistic.py:139]: softplus here as well

File: tensorflow_probability/python/distributions/loglogistic.py [hunk 16]
```diff
@@ -0,0 +1,161 @@
...
+
+  def _log_survival_function(self, x):
+    return - self._log_z(x) + self._log_cdf(x)
```
Comment (srvasude) [inline tensorflow_probability/python/distributions/loglogistic.py:142]: nit: remove leading - sign

File: tensorflow_probability/python/distributions/loglogistic.py [hunk 17]
```diff
@@ -0,0 +1,161 @@
...
+        ]):
+      variance = tf.math.exp(2 * self.loc) * (
+          1. / sinc(2 * scale) - 1. / sinc(scale) ** 2)
+      return tf.where(scale > 0.5, np.nan, variance)
+
+  def _mode(self):
+    log_mode = self.loc + self.scale * (
+        tf.math.log1p(- self.scale) - tf.math.log1p(self.scale)
+    )
+    return tf.where(self.scale > 1., 0., tf.math.exp(log_mode))
```
Comment (srvasude) [inline tensorflow_probability/python/distributions/loglogistic.py:122]: You'll need to cast `0.` and `np.nan` in these wheres unfortunately (this is for dtype=tf.float64).

File: tensorflow_probability/python/distributions/loglogistic_test.py [hunk 18]
```diff
@@ -0,0 +1,209 @@
...
+
+    self.assertAllClose(
+      self.evaluate(pdf),
+      stats.fisk.pdf(x, loc=0., scale=log_logistic_scale, c=1./scale)
+    )
+
+  def testLogLogisticLogPDF(self):
+    log_logistic_scale = 1.5
+    loc = np.log(log_logistic_scale)
+    scale = 1./0.4
```
Comment (srvasude) [inline tensorflow_probability/python/distributions/loglogistic_test.py:141]: Write as 2.5? Here and elsewhere

File: tensorflow_probability/python/distributions/loglogistic_test.py [hunk 19]
```diff
@@ -0,0 +1,209 @@
...
+                        self.evaluate(dist.mean()),
+                        atol=0.1)
+    self.assertAllClose(np.std(samples),
+                        self.evaluate(dist.stddev()),
+                        atol=0.5)
+
+  def testLogLogisticPDF(self):
+    log_logistic_scale = 1.5
+    loc = np.log(log_logistic_scale)
+    scale = 1./0.4
```
Comment (srvasude) [inline tensorflow_probability/python/distributions/loglogistic_test.py:126]: Use batches and broadcasting for these pdf/cdf related tests? Could also add separate tests where you have the parameters loc and scale broadcast, so you have a non-trivial batch of LogLogistic distributions.

**Comment by srvasude** [comment]
Can you address the linter issues mentioned in Travis? Once those are fixed, I can approve and pull in. Thanks again for this PR!

**Comment by npfp** [comment]
> Can you address the linter issues mentioned in Travis? Once those are fixed, I can approve and pull in. Thanks again for this PR!

Done

Many thanks for your review, I learnt a lot!
